docs/SequenceType.md

```diff
@@ -0,0 +1,318 @@
+elements as well as the sequence itself. Currently, the KrnlSeqDeallocOp is used for the
+deallocation and it will be lowered to scf and memref after deallocation pass.
+
+### Load an element from a sequence
+A memref.load could be used to load an element from a sequence. But the loaded memref may
+have a life span longer than the sequence itself, and will have dangling data pointer after
+the sequence has been freed.
+
+To overcome this issue, KrnlSeqExtractOp is introduced. This Op will use memref.load to
+load the element, then use allocate a new memref and copy the data, and finally return
+the copied memref.  This Op is marked with allocation interface and the deallocation pass will insert deallocation for the returned memref automatically. 
+
+### Construct a new sequence from an old sequence
+The sequence ops, SequenceInsert and SequenceErase, construct a new sequence
+with the elements from an old sequence. It is fine to use KrnlSeqExtractOp to get an
+element from the old sequence and use KrnlSeqStoreOp to store it into the new sequence.
+But there are two copy operations for the element. Since it is known that the loaded
+element is only used here while the old sequence is still live, a regular memref.load 
+can be used to avoid one copy. This is a simple optimization for sequence lowering.
+
+## Example
+
+Original .mlir code:
+
+```
+func.func @test_sequence_insert(%arg0: !onnx.Seq<tensor<?x4x5xf32>>, %arg1:tensor<3x4x5xf32>) -> tensor<3xi64>  {
+  %0 = "onnx.Constant"() {value = dense<2> : tensor<1xi64>} : () -> tensor<i64>
+  %1 = "onnx.Add"(%arg1, %arg1) : (tensor<3x4x5xf32>, tensor<3x4x5xf32>) -> tensor<3x4x5xf32>
+  %2 = "onnx.NoValue"() {value} : () -> none
+  %6 = "onnx.SequenceInsert"(%arg0, %1, %2) : (!onnx.Seq<tensor<?x4x5xf32>>, tensor<3x4x5xf32>, none) -> !onnx.Seq<tensor<?x4x5xf32>>
+  %4 = "onnx.SequenceAt"(%6, %0) : (!onnx.Seq<tensor<?x4x5xf32>>, tensor<i64>) -> tensor<?x4x5xf32>
+  %5 = "onnx.Shape"(%4) : (tensor<?x4x5xf32>) -> tensor<3xi64>
+  return %5 : tensor<3xi64>
+}
+```
+
+After --convert-onnx-to-krnl pass
+
+```
+  func.func @test_sequence_insert(%arg0: memref<?xmemref<?x4x5xf32>>, %arg1: memref<3x4x5xf32>) -> memref<3xi64> {
+
+    // onnx.Add
+    %1 = memref.alloc() {alignment = 16 : i64} : memref<3x4x5xf32>
+    %2:3 = krnl.define_loops 3
+    krnl.iterate(%2#0, %2#1, %2#2) with (%2#0 -> %arg2 = 0 to 3, %2#1 -> %arg3 = 0 to 4, %2#2 -> %arg4 = 0 to 5){
+      %22:3 = krnl.get_induction_var_value(%2#0, %2#1, %2#2) : (!krnl.loop, !krnl.loop, !krnl.loop) -> (index, index, index)
+      ...
+      %23 = krnl.load %arg1[%22#0, %22#1, %22#2] : memref<3x4x5xf32>
+      %24 = krnl.load %arg1[%22#0, %22#1, %22#2] : memref<3x4x5xf32>
+      %25 = arith.addf %23, %24 : f32
+      krnl.store %25, %1[%22#0, %22#1, %22#2] : memref<3x4x5xf32>
+    }
+
+    // Sequence Insert
+    %6 = "krnl.seqalloc"(%5) : (index) -> memref<?xmemref<?x4x5xf32>>
```

Comment:
%4, %5 are not defined here. Think the examples does not really need the add, you could just insert arg1 directly in the list arg0.

Reply:
I did not include the whole output.  %4 and %5 is the result of dim of op. The onnx.add is used to show the deallocate of sequence element. If it is an argument of function, no memref.dealloc will be added